agmt/main.py

```
@@ -1082,6 +1090,84 @@ def getTokenLists(sourceId, book):
 	# print(jsonOut)
 	return jsonOut
 
+@app.route("/v1/tokentranslationlist/<projectId>/<book>", methods=["GET"])
+@check_token
+def getTokenTranslationList(projectId, book):
+	'''
+	This method is used to fetch all tokens in a selected book,
+	associated with a project. If a translation is added for it, in that project,
+	that will also be fetched.
+	AUTH: Only accessible to a user who has been assigned that book in that project.
+	INPUT: projectId and 3 letter bookCode(lower or upper cases accepted)
+	OUTPUT FORMAT:
+	[
+	 ['token', null, null],
+	 ['token', 'translation', null],
+	 ['token', 'translation', 'senses1'],
+	 ['token', 'translation', 'senses1,sense2'],
+	 ['token', 'translation', null],
+	 ['token', null, null],
+	 ...
+	]
+	This format of array of arrays and comma seperated string for senses was chosen due to
+	the requirement from UI side (the xlsx npm module)
+	'''
+
+	try:
+		connection = get_db()
+		cursor = connection.cursor()
+		projectId = int(projectId)
+
+		email = request.email
+		cursor.execute("select user_id from autographamt_users where email_id=%s", (email,))
+		userId = cursor.fetchone()[0]
+		cursor.execute("select books from autographamt_assignments where user_id=%s and \
+			project_id=%s", (userId, projectId))
+		assignments = cursor.fetchone()
+		if (not assignments) or (book.lower() not in assignments.split('|')):
+			return '{"success":false, "message":"UnAuthorized! You haven\'t been assigned this book/project"}'
+
+		cursor.execute("select s.table_name from sources as s join autographamt_projects as p \
+		on s.source_id = p.source_id where p.project_id=%s", (projectId,))
+		source_table = cursor.fetchone()[0]
+		tablename = source_table + '_tokens'
+
+		cursor.execute("select book_id from bible_books_look_up where book_code=%s", (book,))
```

You can add a check to see if the book is valid by checking if its in the bible_books_look_up table, in case it is not you can return a error, you can do it before line 1127 where book is used. if it is not valid the query result will be empty

---

Added